Optee docs refactoring

jbech-linaro opened the pull request:
Majority of the doc cleanup has been done here. I'm about to create a PR in build.git also, where the stuff will be added instead.

Even though I cleaned up optee_design.md I still think it would be good to revise the information in more detail as soon as possible. I also think that @etienne-lms changes related to shared memory for SDP should be included there. But .. first things first.

**Edit**
Counterpart in build.git: https://github.com/OP-TEE/build/pull/94

jforissier commented on README.md:
```diff
@@ -769,15 +87,9 @@ the code also follows GlobalPlatform standards. The exceptions are as follows:
-an environment variable, namely CHECKPATCH. So, suppose that the source code for
-the Linux kernel is at `$HOME/devel/linux`, then you have to export like follows:
+cannot include this script directly into this project. Please use checkpatch
+directly from the Linux kernel git.
```
Right, and you may mention the helper script: `scripts/checkpatch.sh`

jforissier commented:
For the 3 "Remove" patches:
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -327,69 +169,62 @@ Different stacks are used during different stages. The stacks are:
-temporary storage of registers and also keeps the relevant stack pointers.
-In general when we talk about assigning a stack pointer to the CPU below we
-mean `SP_EL0`
+There's only two stack pointers, `SP_EL1` and `SP_EL0`, available for OP-TEE in
```
s/There's/There are/

jbech-linaro replied:
Done

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -327,69 +169,62 @@ Different stacks are used during different stages. The stacks are:
+AArch64. When an exception is received stack pointer is always `SP_EL1` which is
+used temporarily while assigning an appropriate stack pointer for `SP_EL0`.
+**`SP_EL1` is always assigned the value of `thread_core_local[cpu_id]`.** This
+structure have some spare space for temporary storage of registers and also
```
s/have/has/

jbech-linaro replied:
Done

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -327,69 +169,62 @@ Different stacks are used during different stages. The stacks are:
-that saves the thread state instead. The thread is resumed in the same way
-though.
+IRQ exit occurs when OP-TEE receives an IRQ, which is always handled in normal
+world. IRQ exit is similar to RPC exit but it's thread_irq_handler() that saves
```
thread_irq_handler() and elx_irq() (respectively for ARMv7/Aarch32 and for Aarch64)

jbech-linaro replied:
Done

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -404,180 +239,165 @@ thread_svc_regs` in case the TA is executed in AArch64.
+
+Note that:
+- The shared memory pool is physically contiguous.
+- The shared memory area is not protected as it is used by both non-secure and
```
s/not protected/not secure/

jbech-linaro replied:
Done

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -404,180 +239,165 @@ thread_svc_regs` in case the TA is executed in AArch64.
+configuration for the platform of interest. The Linux driver will then
+initialize the shared memory pool accordingly.
 
 ### Shared Memory Chunk Allocation
```
this section should be updated: shm is allocated though linux kernel generic pool support (CONFIG_GENERIC_ALLOCATOR).

jbech-linaro replied:
I agree, that was a part of what I had in mind with updates in general regarding changes to shared memory lately. There is both the work you've done (the extensions) and then @lorc did some work just recently.

etienne-lms replied:
As of optee-v2.3.0, i think we should replace this full section `### Shared Memory Chunk Allocation` with proposal below. Once the new shm handling from @lorc and @jenswi-linaro is merge, we will update with appropriate description.

It is the Linux kernel driver for OP-TEE that is responsible for allocating chunks of shared memory.  OP-TEE linux kernel driver relies on linux kernel generic allocation support (CONFIG_GENERIC_ALLOCATION) to allocation/release of shared memory physical chunks. OP-TEE linux kernel driver relies on linux kernel dma-buf support (CONFIG_DMA_SHARED_BUFFER) to track shared memory buffers references.

jbech-linaro replied:
I've replaced the text according to your suggestion here.

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -404,180 +239,165 @@ thread_svc_regs` in case the TA is executed in AArch64.
+just a few KiB of SRAM can’t be expected to be able to run a complete TEE
+solution in SRAM. But those with 128 to 256 KiB of SRAM can be expected to have
+a capable TEE solution in SRAM. The pager provides a solution to this by demand
+paging readonly parts of OP-TEE using virtual memory.
```
remove "readonly parts".

jbech-linaro replied:
Done

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -404,180 +239,165 @@ thread_svc_regs` in case the TA is executed in AArch64.
-itself is not encrypted.
+TrustZone protected DRAM or in some cases non-secure DRAM is used as backing
+store. The data in the backing store is integrity protected with one hash
+(SHA-256) per page (4KiB). Only readonly pages are kept in backing store for
```
remove text from "Only readonly pages..." to "...need to be encrypted too."

jbech-linaro replied:
Done

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -404,180 +239,165 @@ thread_svc_regs` in case the TA is executed in AArch64.
+store. The data in the backing store is integrity protected with one hash
+(SHA-256) per page (4KiB). Only readonly pages are kept in backing store for
+performance reasons as readwrite data would need to be encrypted too. Readonly
+pages doesn't need to be encrypted since the OP-TEE binary itself is not
```
s/doesn't need to be/are not/

jbech-linaro replied:
Done

etienne-lms commented on documentation/optee_design.md:
````diff
@@ -404,180 +239,165 @@ thread_svc_regs` in case the TA is executed in AArch64.
++--------------+-----------------+
 ```
-Where "nozi" stands for "not zero initialized", this section contains small
+Where "`nozi`" stands for "not zero initialized", this section contains small
````
s/small stacks/entry stacks/
Add "thread stack (when TEE pager is disable)".
Add "TEE pager cached translation table (in case pager is enable and LPAE MMU is used)".

jbech-linaro replied:
Fixed, but please read my changes to see if that is in line with what you meant to say.

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -611,68 +431,68 @@ struct optee_header {
+`(init_load_addr_hi << 32 | init_load_addr_lo)`. `init_mem_usage` is used by the
+loader to be able to check that there's enough physical memory available for
+OP-TEE to be able to initialize at all. The loader supplies the address of the
+first byte following what wasn't copied in r0/x0 and jumps to the load address
```
-The loader supplies the address of the first byte following what wasn't copied in r0/x0 and jumps
+The loader supplies in r0/x0 the address of the first byte following what wasn't copied and jumps

jbech-linaro replied:
Done

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -611,68 +431,68 @@ struct optee_header {
+The pager is initialized as early as possible during boot in order to minimize
+the "init" area. As a consequence of this the old initialization has been broken
+up a bit since the pager initialization depends on some internal APIs being
+available. The global variable `tee_mm_vcore` describes the virtual memory range
```
Remove "As a consequence of this the old initialization has been broken up a bit since the pager initialization depends on some internal APIs being available."

jbech-linaro replied:
Done

etienne-lms commented on documentation/optee_design.md:
````diff
@@ -611,68 +431,68 @@ struct optee_header {
-call to `tee_pager_add_area()`.
-```
+A virtual memory range to be handled by the pager is registered with a call to
+`tee_pager_add_area()`.
````
s/tee_pager_add_area/tee_pager_add_core_area/
Description below must be update...

jbech-linaro replied:
Done

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -611,68 +431,68 @@ struct optee_header {
-The pages in the "init" area are supplied with unmap == false since those
-page have valid content and are in use.
+`tee_pager_add_pages()` expects all physical memory to be mapped in the level 2
+translation table already. `tee_pager_add_pages()` takes physical address stored
```
Remove "`tee_pager_add_pages()` expects all physical memory to be mapped in the level 2 translation table already".

jbech-linaro replied:
Done

jbech-linaro commented on documentation/optee_design.md:
Done

jbech-linaro commented on documentation/optee_design.md:
Done

jbech-linaro commented on documentation/optee_design.md:
Done, rewrote it slightly.

etienne-lms commented:
Related to my review tag, i can only acknowledge the removed part due to dependency on https://github.com/OP-TEE/build/pull/94.

jbech-linaro commented on README.md:
````diff
@@ -91,668 +65,12 @@ platforms have different sub-maintainers, please refer to the file
-```
-
----
-### 4.6 Freescale MX6UL EVK
````
@MrVan , we have to discuss how to deal with this section. All other platforms (actively maintained) have a repo setup. Do you think that is something you could add too? Or is that to far away from the use cases for the Freescale devices.

MrVan replied:
@jbech-linaro I'll add a repo for this board. Only need to post patches to `https://github.com/OP-TEE/manifest`?

jbech-linaro replied:
@MrVan , yes, but there is no hurry doing that I think. What I have in the pipe is this change and most of the information removed here went into https://github.com/OP-TEE/build/pull/94. So I was thinking if it was possible in your case to:
a) Add the needed xml-files (manifests) in https://github.com/OP-TEE/manifest
b) If it makes sense, create similar "helper" Makefile as we have in https://github.com/OP-TEE/build. If it doesn't make sense, then we can create a separate file for the device(s) your maintaining in like I've done with all the "Add __xyz__.md" file in the PR (https://github.com/OP-TEE/build/pull/94).

Again, no hurry, since I don't think the updated doc patches will be part of the 2.3.0 release (maybe, maybe not).

etienne-lms commented:
+1

jbech-linaro commented:
Squashed, rebased and tag(s) applied (should be) ready for merge! Thanks!

And regarding MrVan's recent changes for the Freescale platform, let's deal with them separately in build.git (I.e., I plan to at least start by pulling the current text and put it there instead).

jbech-linaro commented:
I can see that it's Jens latest and just merged change that made this conflict. I'll rebase again.
